### DISCUSSION TITLE
Custom ruleset

### DISCUSSION THREAD

**Discussion by alivanni** [pmd/pmd#4761] [Q&A] [answered]
Hi,

I am going trough the process of creating custom ruleset described here: https://docs.pmd-code.org/pmd-doc-6.55.0/pmd_userdocs_making_rulesets.html
What is not clear to me - does custom ruleset initially completely empty in terms of functionality, or it kind of implicitly inherits all the default rules and behaviour? Other words, if I customise one existing rule in my ruleset - this rule is the only thing that going to be checked or default behaviour of all other rules will also be executed?

Thanks,
Aleksandr

**Comment by jsotuyod** [comment]
There are no default rules in PMD (some tools such as Gradle may set defaults on their own, but that's not curated by us).

Having said that, you can provide multiple ruleset files to PMD, so you can provide whatever defaults you have + your custom ruleset customizing that one rule; just be mindful that the order matters (if I recall correctly, first one wins).

So, if you set 2 ruleset files `ruleSet1.xml` and `ruleSet2.xml` in that order, and both configure rule A in a different way, the version from `ruleSet1.xml` is the one used.

This is however undocumented behavior, I don't think we even print a warning for this…

**Reply by alivanni** [reply]
I see. And is it possible instead of maintaining two rulesets somehow inherit all java rules for instance in my xml and then apply customisation on top of that? What complicates the thing is that I am using PMD in a context of Codacy code check. Probably "default" set is somehow defined inside Codacy itself. But now I am quite confuse how can I customise singe Java rule without disabling all other rules.

**Reply by jsotuyod** [reply, marked as answer]
Yes you can! you can have your ruleset import a complete category and exclude / configure any rules within that.

For instance this one here does exactly that, importing all existing categories (except security) and manually excluding / overriding particular rules. https://github.com/Monits/static-code-analysis-plugin/blob/staging/defaults/pmd/pmd-6.xml

**Reply by adangel** [reply]
This is actually documented here: https://docs.pmd-code.org/latest/pmd_userdocs_making_rulesets.html#bulk-adding-rules under "Bulk-adding rules"

**Reply by akansha1808** [reply]
Hye! actually I also tried to create my own custom ruleset but don't know exactly how to override the message for a particular rule in my ruleset. I tried to do in the way it is documented but it still showed the original message. For eg. I tried to modify the message of UseUtilityClass in quickstart.xml ruleset but it didnot change. Can anyone please suggest me a way how i can modify the message of a particular rule.

**Reply by jsotuyod** [reply]
That is actually an implementation issue. The rule implementation hardcodes a message instead of using the on defined in the XML. There are legit uses of this (ie: rules that provide multiple messages with greater detail for particular situations), but this one only produces the one message…

https://github.com/pmd/pmd/blob/ae76c8f29302f71dfd1bbb726e65f73e4f2bcb0b/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/UseUtilityClassRule.java#L81-L82

I created https://github.com/pmd/pmd/issues/5023 to track this issue.

**Reply by akansha1808** [reply]
Thankyou so much!!

**Reply by akansha1808** [reply]
Hello @jsotuyod I came across one more rule whose message is also not getting changed, and I assume it's message is also hardcoded. The rule is "UnnecessaryImportRule", can you also please confirm?

**Reply by oowekyala** [reply]
UnnecessaryImport overrides the default message to provide more detailed messages. It is expected that the message cannot be overridden.

What is your use case for overriding messages? Are you translating them, or are you dissatisfied with the default messages?

**Reply by akansha1808** [reply]
Actually I wanted to provide a more detailed message so was trying to modify the default message with my custom message in my ruleset. Can you please help me in knowing how to change the default message of UnnecessaryImportRule?

**Reply by oowekyala** [reply]
The message is set in code in Java so you will have to modify [the Java source](https://github.com/pmd/pmd/blob/master/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/UnnecessaryImportRule.java) and recompile the module. If you could explain how the message could be more detailed and you think this change could benefit others please open a feature request.

**Reply by akansha1808** [reply]
Thankyou so much for this help...actually I was trying to modify messages and check the feature of pmd, but the message is nothing as such which could be of better help. It was just a part of my exploration.

**Reply by akansha1808** [reply]
Hye! I had a query related to the priority of the rules. Can I have a ruleset containing rules with all the different priorities and will I be able to run pmd test with the same ruleset based on a specific priority for eg. I would like that only the rules having priority 1 should show up. Is there any way for that??

**Reply by jsotuyod** [reply]
Yes there is! It will obviously depend on how you run PMD, but the docs are complete.

[Here is the reference](https://docs.pmd-code.org/latest/pmd_userdocs_cli_reference.html#-minimum-priority) for doing this in the CLI.

**Reply by akansha1808** [reply]
Hye @jsotuyod while exploring the codebase of pmd, I was trying to work around with the code and in the course I tried to change the name of the cli comands. For eg. I tried to change the command pmd --version to pmd --feature, so for that I changed the text in the class DesignerCommand.java, but even after changing the text from "version" to "feature", the command didnot change. Can you please help me know how can I change the command?

**Reply by oowekyala** [reply]
DesignerCommand is the class that defines options for the command `pmd designer`. You might have been renaming `pmd designer --version` to `--feature`

**Reply by jsotuyod** [reply]
The standard `--version` flag is managed by PicoCli (the library in use), we only hook up the version provider to create the appropriate output

https://github.com/pmd/pmd/blob/28c6c8e1210e1c4591ba8aa1d0ae537a538eb104/pmd-cli/src/main/java/net/sourceforge/pmd/cli/commands/internal/PmdRootCommand.java#L11

Each subcommand (`pmd *check*`, `pmd *cpd*`, `pmd *designer*`, etc.) each has it's own class. The one for PMD analysis is [`PmdCommand`](https://github.com/pmd/pmd/blob/master/pmd-cli/src/main/java/net/sourceforge/pmd/cli/commands/internal/PmdCommand.java). I recommend you check the [PicoCli documentation](https://picocli.info/) if you want to play with it.
